mapper/cumulocity/c8y_translator_lib/src/serializer.rs

```diff
@@ -0,0 +1,128 @@
+use chrono::prelude::*;
+use std::io::Write;
+use thin_edge_json::{json::ThinEdgeJsonError, measurement::GroupedMeasurementVisitor};
+
+pub struct C8yJsonSerializer {
+    buffer: Vec<u8>,
+    is_within_group: bool,
+    needs_separator: bool,
+    timestamp_present: bool,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum C8yJsonSerializationError {
+    #[error(transparent)]
+    IoError(#[from] std::io::Error),
+
+    #[error(transparent)]
+    MeasurementCollectorError(#[from] MeasurementStreamError),
+
+    #[error(transparent)]
+    ThinEdgeJsonParseError(#[from] ThinEdgeJsonError),
+}
+#[derive(thiserror::Error, Debug, Eq, PartialEq)]
+pub enum MeasurementStreamError {
+    #[error("Unexpected time stamp within a group")]
+    UnexpectedTimestamp,
+
+    #[error("Unexpected end of data")]
+    UnexpectedEndOfData,
+
+    #[error("Unexpected end of group")]
+    UnexpectedEndOfGroup,
+
+    #[error("Unexpected start of group")]
+    UnexpectedStartOfGroup,
+}
+
```

Comment:
Why do you need to have 2 separate error types here? Can't you combine them into one and use as such?

Reply:
I like they are separated. The highest hierarchy is `C8yJsonSerializationError`, then we have a set of collection errors under the top hierarchy.